concept/type_/annotation.rs

```diff
@@ -82,11 +105,118 @@ impl AnnotationRegex {
         Self { regex }
     }
 
+    pub fn default() -> Self {
+        Self::new(".*".to_owned())
+    }
+
     pub fn regex(&self) -> &str {
         &self.regex
     }
 }
 
+impl Annotation {
+    pub fn category(&self) -> AnnotationCategory {
+        match self {
+            Self::Abstract(_) => AnnotationCategory::Abstract,
+            Self::Distinct(_) => AnnotationCategory::Distinct,
+            Self::Independent(_) => AnnotationCategory::Independent,
+            Self::Unique(_) => AnnotationCategory::Unique,
+            Self::Key(_) => AnnotationCategory::Key,
+            Self::Cardinality(_) => AnnotationCategory::Cardinality,
+            Self::Regex(_) => AnnotationCategory::Regex,
+            Self::Cascade(_) => AnnotationCategory::Cascade,
+        }
+    }
+}
+
+#[derive(Debug, Default, Copy, Clone, Eq, PartialEq, Hash)]
+pub struct AnnotationCascade;
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum AnnotationCategory {
+    Abstract,
+    Distinct,
+    Independent,
+    Unique,
+    Key,
+    Cardinality,
+    Regex,
+    Cascade,
+    // TODO: Subkey
+    // TODO: Values
+    // TODO: Range
+    // TODO: Replace
+}
+
+impl AnnotationCategory {
+    pub fn to_default_annotation(&self) -> Annotation {
+        // TODO: fn const for regex default
+        match self {
+            AnnotationCategory::Abstract => Annotation::Abstract(AnnotationAbstract),
+            AnnotationCategory::Distinct => Annotation::Distinct(AnnotationDistinct),
+            AnnotationCategory::Independent => Annotation::Independent(AnnotationIndependent),
+            AnnotationCategory::Unique => Annotation::Unique(AnnotationUnique),
+            AnnotationCategory::Key => Annotation::Key(AnnotationKey),
+            AnnotationCategory::Cardinality => Annotation::Cardinality(AnnotationCardinality::default()),
+            AnnotationCategory::Regex => Annotation::Regex(AnnotationRegex::default()),
+            AnnotationCategory::Cascade => Annotation::Cascade(AnnotationCascade),
+        }
+    }
+
+    pub fn compatible_to_declare_together(&self, other: AnnotationCategory) -> bool {
```

Comment:
Decided to move annotation checks here, as it feels strange to have **that** many annotation validations outside of annotations. Plus it complicated naming in many cases. 

This function is to check that two annotations can be declared together on one *thing*, being neighbors.

Reply:
I like `category.declarable_alongside(other_category)`, but not sure about the second one. For context:
* I use this function (`compatible_to_transitively_add`) both to check that my newly declared annotation is compatible with inherited annotations (if I have inherited `@key`, I cannot declare `@unique` or `@card`) AND to filter out annotations when traversing types from a *subtype* to the root, collecting all the transitive annotations for this *subtype* (if I already collected `@key` from *subtype* (or its parents), I don't need to add `@unique` or `@card` from their parents on top).
* We have `declared` now as a term for annotations that are directly connected to a *thing*, and using this term for transitive compatibility seems a bit confusing
* `above` and `below` do not work here either because it's kinda natural to use this function in both directions on different stages (validation and collection)

Reply:
I'll leave both functions as is for now to preserve symmetry in their names, will be ready to change both after we finish our disc